source/agora/node/FullNode.d

```diff
@@ -300,16 +300,40 @@ public class FullNode : API
                     return;
 
                 this.network.getBlocksFrom(
-                    this.ledger.getBlockHeight() + 1,
-                    // if any blocks fail validation => short-circuit
-                    blocks => blocks.all!(block => this.acceptBlock(block)));
+                    Height(this.ledger.getBlockHeight() + 1),
+                    &addBlocks);
                 this.network.getUnknownTXs(this.ledger);
             }
             catchup(); // avoid delay
             this.taskman.setTimer(this.network.node_config.block_catchup_interval, &catchup, Periodic.Yes);
         });
     }
 
+    /***************************************************************************
+
+        Add blocks to the ledger and add the related pre-images
+
+        Params:
+            blocks = the blocks to be added
+            preimages = the preimages needed to check the validity of the blocks
+
+        Returns:
+            true if the blocks and preimages was added
+
+    ***************************************************************************/
+
+    bool addBlocks(const(Block)[] blocks, const(PreImageInfo)[] preimages) @safe
+    {
+        foreach (block; blocks)
+        {
+            if(!this.ledger.addPreimages(preimages))
```

Comment:
Instead of adding a function in the ledger whose full role is to forward to the enrollment manager, call the enrollment manager directly and remove the Ledger function